lib/p2p/Peer.ts

```diff
@@ -131,7 +133,14 @@ class Peer extends EventEmitter {
 
     await this.initConnection();
     this.initStall();
-    await this.initHello(handshakeData);
+    this.sendHello(handshakeData);
+    if (this.handshakePromise) {
+      // we were already waiting for an incoming hello packet, finish waiting before opening
+      await this.handshakePromise;
+    } else {
+      // we're not already waiting for an incoming hello packet, so wait now
```

Comment:
why this condition would ever occur, if you're waiting for the packet as part of `initConnection`? this makes the previous `initHello` pretty much the same, just that you moved the `this.wait` call to the connection phase.

Reply:
It should never occur, although I put a check there just in case. I could just put a `!` operator on `handshakePromise` or maybe still have the check but close the peer if we're not waiting for a handshake at this point for some unexpected reason.

Reply:
I have a cleaner approach almost ready which accepts hello packets at any time (potentially subject to some anti-spam limits down the road) to set or update the handshake state, I'll update the PR tomorrow.